tests/phpunit/tests/rest-api/wpRestBlockPatternCategoriesController.php

```diff
@@ -0,0 +1,145 @@
+ *
+ * @group restapi
+ */
+class Tests_REST_WpRestBlockPatternCategoriesController extends WP_Test_REST_Controller_Testcase {
+
+	/**
+	 * Admin user ID.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @var int
+	 */
+	protected static $admin_id;
+
+	/**
+	 * Original instance of WP_Block_Patterns_Registry.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @var WP_Block_Patterns_Registry
+	 */
+	protected static $orig_registry;
+
+	/**
+	 * Instance of the reflected `instance` property.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @var ReflectionProperty
+	 */
+	private static $registry_instance_property;
+
+	/**
+	 * The REST API route.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @var string
+	 */
+	const REQUEST_ROUTE = '/wp/v2/block-patterns/categories';
+
+	/**
+	 * Set up class test fixtures.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param WP_UnitTest_Factory $factory WordPress unit test factory.
+	 */
+	public static function wpSetupBeforeClass( $factory ) {
+		self::$admin_id = $factory->user->create( array( 'role' => 'administrator' ) );
+
+		// Setup an empty testing instance of `WP_Block_Pattern_Categories_Registry` and save the original.
+		self::$orig_registry              = WP_Block_Pattern_Categories_Registry::get_instance();
+		self::$registry_instance_property = new ReflectionProperty( 'WP_Block_Pattern_Categories_Registry', 'instance' );
+		self::$registry_instance_property->setAccessible( true );
+		$test_registry = new WP_Block_Pattern_Categories_Registry();
+		self::$registry_instance_property->setValue( $test_registry );
+
+		// Register some categories in the test registry.
+		$test_registry->register( 'test', array( 'label' => 'Test' ) );
+		$test_registry->register( 'query', array( 'label' => 'Query' ) );
+	}
+
+	public static function wpTearDownAfterClass() {
+		self::delete_user( self::$admin_id );
+
+		// Restore the original registry instance.
+		self::$registry_instance_property->setValue( self::$orig_registry );
+		self::$registry_instance_property->setAccessible( false );
+		self::$registry_instance_property = null;
+		self::$orig_registry              = null;
+	}
+
+	public function set_up() {
+		parent::set_up();
+
+		switch_theme( 'emptytheme' );
+	}
+
+	public function test_register_routes() {
+		$routes = rest_get_server()->get_routes();
+		$this->assertArrayHasKey( static::REQUEST_ROUTE, $routes );
+	}
+
+	public function test_get_items() {
+		wp_set_current_user( self::$admin_id );
+
+		$expected_names  = array( 'test', 'query' );
+		$expected_fields = array( 'name', 'label' );
+
+		$request            = new WP_REST_Request( 'GET', static::REQUEST_ROUTE );
+		$request['_fields'] = 'name,label';
+		$response           = rest_get_server()->dispatch( $request );
+		$data               = $response->get_data();
+
+		$this->assertCount( count( $expected_names ), $data );
+		foreach ( $data as $idx => $item ) {
+			$this->assertSame( $expected_names[ $idx ], $item['name'] );
+			$this->assertSame( $expected_fields, array_keys( $item ) );
+		}
+	}
```

Comment:
I would like to see a test that verifies the capabilitiy checks are working properly. Typically we include a logged-out user, and a user with subscriber level capabilities.

Reply:
✅ See f7ce9521dc.